library/repair_docker_data_volumes

```diff
@@ -0,0 +1,338 @@
+      - should we fix the VFS links (issue 1)?
+    requried: true
+  fix_container_mountpoints:
+    description:
+      - should we fix the container mountpoints (issue 2)?
+    required: true
+  old_container_configs:
+    description:
+      - config data from containers, before the package update
+'''
+
+CONTAINER_ROOT_DIR = '/var/lib/docker/containers'
+CONFIG_V2_FILENAME = 'config.v2.json'
+CONFIG_V3_FILENAME = 'config.v3.json'
+BACKUP_FILENAME = 'config.v2.json.pre_volume_fix'
+VFS_BASE_DIR = '/var/lib/docker/vfs/dir'
+VOLUMES_ROOT_DIR = '/var/lib/docker/volumes'
+LOCAL_DRIVER = 'local'
+
+VFS_MOUNT_REGEX = re.compile(VFS_BASE_DIR + "/([a-f0-9]+)$")
+VOLUME_MOUNT_REGEX = re.compile(VOLUMES_ROOT_DIR + "/([a-f0-9]+)/_data")
+
+
+class Volume(object):
+    def __init__(self, driver, name):
+        self.driver = driver
+        self.name = name
+
+    def inspect(self, module):
+        try:
+            inspect_output = subprocess.check_output(
+                ['docker', 'volume', 'inspect', self.name],
+                stderr=subprocess.STDOUT)
+            return json.loads(inspect_output)[0]
+        except subprocess.CalledProcessError, c:
+            module.fail_json(
+                msg="failure while inspecting volume {}".format(self.name),
+                exception=str(c))
+
+    @property
+    def is_local(self):
+        return self.driver == LOCAL_DRIVER
+
+    @property
+    def vfs_path(self):
+        return os.path.join(VFS_BASE_DIR, self.name)
+
+
+def _volumes(module):
+    try:
+        volumes_output = subprocess.check_output(
+            ['docker', 'volume', 'ls'],
+            stderr=subprocess.STDOUT)
+
+        return [
+            Volume(driver=line.split()[0], name=line.split()[1])
+            for line in volumes_output.split('\n')[1:] if line != '']
+
+    except subprocess.CalledProcessError, c:
+        module.fail_json(
+            msg="failure while listing volumes",
+            exception=str(c))
+
+
+class CreateSymlinkOperation(object):
+    def __init__(self, source_path, link_path):
+        self.source_path = source_path
+        self.link_path = link_path
+
+    def run(self, module):
+        subprocess.check_call(['ln', '-sfn', self.source_path, self.link_path])
+
+    def desc(self):
+        return "creating symlink from {} to {}".format(self.source_path, self.link_path)
+
+
+def fix_volume_mount_missing_symlinks(module):
+    for volume in _volumes(module):
+        if not volume.is_local:
+            # don't deal with volumes that don't use the "local" driver.
+            continue
+
+        inspect_data = volume.inspect(module)
+        if os.path.exists(inspect_data['Mountpoint']):
+            # mount point for this volume is non-broken
+            continue
+
+        if not os.path.exists(volume.vfs_path):
+            # mount data isn't where we expect it to be
+            continue
+
+        yield CreateSymlinkOperation(volume.vfs_path, inspect_data['Mountpoint'])
```

Comment:
I think the sense of the tests and the comments about them are much clearer if you use an explicit generator expression here than with the generator function created by using `yield`.

```
def fix_volume_mount_missing_symlinks(module):
    return (
        CreateSymlinkOperation(volume.vfs_path, inspect_data['Mountpoint'])
        for volume in _volumes(module)

        # deal only with volumes that use the "local" driver
        if volume.is_local

        for inspect_data in [volume.inspect(module)]

        # mount point for this volume is broken
        if not os.path.exists(inspect_data['Mountpoint'])

        # mount data is where we expect it to be
        if os.path.exists(volume.vfs_path)
    )
```
`for inspect_data in [volume.inspect(module)]` is the only thing that feels hacky. I wrote it that way so we get the named `inspect_data` for the singleton member of the list.